db/chained_delegating_reader.hh

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2020 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <seastar/core/shared_future.hh>
+
+#include "flat_mutation_reader.hh"
+
+// A reader which allows to insert a deferring operation before reading.
+// All calls will first wait for a future to resolve, then forward to a given underlying reader.
+class chained_delegating_reader : public flat_mutation_reader::impl {
+    std::unique_ptr<flat_mutation_reader> _underlying;
+    std::function<future<flat_mutation_reader>()> _populate_reader;
+    shared_future<> _ready;
+    std::function<void()> _on_destroyed;
+    bool _populating_started;
+    
+public:
+    chained_delegating_reader(schema_ptr s, std::function<future<flat_mutation_reader>()> populate, reader_permit permit, std::function<void()> on_destroyed = []{})
+        : impl(s, std::move(permit))
+        , _populate_reader(std::move(populate))
+        , _on_destroyed(std::move(on_destroyed))
+        , _populating_started(false)
+    { }
+
+    chained_delegating_reader(chained_delegating_reader&& rd) = delete;
+
+    ~chained_delegating_reader() {
+        _on_destroyed();
+    }
+
+    virtual future<> fill_buffer(db::timeout_clock::time_point timeout) override {
+        if (!_populating_started) {
+            _populating_started = true;
+            _ready = _populate_reader().then([this] (flat_mutation_reader&& rd) {
+                _underlying = std::make_unique<flat_mutation_reader>(std::move(rd));
+            });
+        }
+
+        if (!_ready.available()) {
+            return _ready.get_future(timeout).then([this, timeout] {
```

Comment:
If this times out, who waits for _populate_reader() to complete? If this reader is destroyed while _populate_reader is running, we will have use-after-free.

The simplest way out is to not timeout during population, or propagate the timeout to _populate_reader.

Reply:
_ready is not necessary then.

Reply:
Is it ok to not involve timeouts in the population if the populate function can have basically anything inside? Like in our case, we execute the mutations.

Reply:
It's better to propagate a timeout.